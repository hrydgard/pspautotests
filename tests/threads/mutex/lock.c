#include "shared.h"
#include <limits.h>

// Test sceKernelLockMutex with an exising mutex.
#define LOCK_TEST_SIMPLE(title, mutex, count) { \
	int result = sceKernelLockMutex(mutex, count, NULL); \
	if (result == 0) { \
		schedf("%s: ", title); \
		schedfMutex(mutex); \
	} else { \
		schedf("%s: Failed (%X)\n", title, result); \
	} \
}

// Test sceKernelLockMutex with NULL timeout.
#define LOCK_TEST(title, attr, initial, count) { \
	SceUID mutex = sceKernelCreateMutex("lock", attr, initial, NULL); \
	LOCK_TEST_SIMPLE(title, mutex, count); \
	sceKernelDeleteMutex(mutex); \
}

// Test sceKernelLockMutex with a timeout.
#define LOCK_TEST_TIMEOUT(title, attr, initial, count, initial_timeout) { \
	SceUID mutex = sceKernelCreateMutex("lock", attr, initial, NULL); \
	SceUInt timeout = initial_timeout; \
	int result = sceKernelLockMutex(mutex, count, &timeout); \
	if (result == 0) { \
		schedf("%s: OK (%dms left)\n", title, timeout); \
		schedfMutex(mutex); \
	} else { \
		schedf("%s: Failed (%X, %dms left)\n", title, result, timeout); \
	} \
	sceKernelDeleteMutex(mutex); \
}

// Test sceKernelLockMutex by locking on a separate thread first.
#define LOCK_TEST_TIMEOUT_THREAD(title, attr, initial, count, initial_timeout) { \
	schedf("%s: ", title); \
	schedulingResult = -1; \
	SceUID mutex = sceKernelCreateMutex("lock", attr, initial, NULL); \
	sceKernelStartThread(lockThread, sizeof(int), &mutex); \
	sceKernelDelayThread(400); \
	SceUInt timeout = initial_timeout; \
	int result = sceKernelLockMutex(mutex, count, &timeout); \
	schedf("L2 "); \
	sceKernelDelayThread(600); \
	sceKernelDeleteMutex(mutex); \
	sceKernelWaitThreadEnd(lockThread, NULL); \
	if (result == 0) { \
		schedf("OK (thread=%08X, %dms left)\n", schedulingResult, timeout); \
	} else { \
		schedf("Failed (thread=%08X, main=%08X, %dms left)\n", schedulingResult, result, timeout); \
	} \
	sceKernelTerminateThread(lockThread); \
}

static int lockFunc(SceSize argSize, void* argPointer) {
	SceUInt timeout = 1000;
	schedulingResult = sceKernelLockMutex(*(int*) argPointer, 1, &timeout);
	schedf("L1 ");
	sceKernelDelayThread(1000);
	return 0;
}

static int deleteMeFunc(SceSize argSize, void* argPointer) {
	int result = sceKernelLockMutex(*(int*) argPointer, 1, NULL);
	schedf("After delete: %08X\n", result);
	return 0;
}

static int exitFunc(SceSize argSize, void* argPointer) {
	int result = sceKernelLockMutex(*(int*) argPointer, 1, NULL);
	schedf("Now exiting: %08X\n", result);
	return 0;
}

int main(int argc, char **argv) {
	LOCK_TEST("Lock 0 => 0", PSP_MUTEX_ATTR_FIFO, 0, 0);
	LOCK_TEST("Lock 0 => 1", PSP_MUTEX_ATTR_FIFO, 0, 1);
	LOCK_TEST("Lock 0 => 2", PSP_MUTEX_ATTR_FIFO, 0, 2);
	LOCK_TEST("Lock 0 => -1", PSP_MUTEX_ATTR_FIFO, 0, -1);
	LOCK_TEST("Lock 1 => 1", PSP_MUTEX_ATTR_FIFO, 1, 1);
	LOCK_TEST("Lock 0 => 2 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 0, 2);
	LOCK_TEST("Lock 0 => -1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 0, -1);
	LOCK_TEST("Lock 1 => 1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 1, 1);
	LOCK_TEST("Lock 1 => INT_MAX - 1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 1, INT_MAX - 1);
	LOCK_TEST("Lock 1 => INT_MAX (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 1, INT_MAX);
	LOCK_TEST("Lock INT_MAX => INT_MAX (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, INT_MAX, INT_MAX);

	LOCK_TEST_TIMEOUT("Lock 0 => 0", PSP_MUTEX_ATTR_FIFO, 0, 0, 500);
	LOCK_TEST_TIMEOUT("Lock 0 => 1", PSP_MUTEX_ATTR_FIFO, 0, 1, 500);
	LOCK_TEST_TIMEOUT("Lock 0 => 2", PSP_MUTEX_ATTR_FIFO, 0, 2, 500);
	LOCK_TEST_TIMEOUT("Lock 0 => -1", PSP_MUTEX_ATTR_FIFO, 0, -1, 500);
	LOCK_TEST_TIMEOUT("Lock 1 => 1", PSP_MUTEX_ATTR_FIFO, 1, 1, 500);
	LOCK_TEST_TIMEOUT("Lock 0 => 2 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 0, 2, 500);
	LOCK_TEST_TIMEOUT("Lock 0 => -1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 0, -1, 500);
	LOCK_TEST_TIMEOUT("Lock 1 => 1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 1, 1, 500);
	LOCK_TEST_TIMEOUT("Zero timeout", PSP_MUTEX_ATTR_FIFO, 0, 1, 0);

	SceUID lockThread = CREATE_SIMPLE_THREAD(lockFunc);
	LOCK_TEST_TIMEOUT_THREAD("Locked 1 => 1", PSP_MUTEX_ATTR_FIFO, 1, 1, 500);
	LOCK_TEST_TIMEOUT_THREAD("Locked 0 => 1", PSP_MUTEX_ATTR_FIFO, 0, 1, 500);
	LOCK_TEST_TIMEOUT_THREAD("Locked 1 => 1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 1, 1, 500);
	LOCK_TEST_TIMEOUT_THREAD("Locked 0 => 1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 0, 1, 500);
	LOCK_TEST_TIMEOUT_THREAD("Locked zero timeout", PSP_MUTEX_ATTR_FIFO, 0, 1, 0);

	SceUID deleteThread = CREATE_SIMPLE_THREAD(deleteMeFunc);
	SceUID mutex = sceKernelCreateMutex("lock", 0, 1, NULL);
	sceKernelStartThread(deleteThread, sizeof(int), &mutex);
	sceKernelDelayThread(500);
	sceKernelDeleteMutex(mutex);

	LOCK_TEST_SIMPLE("NULL => 0", 0, 0);
	LOCK_TEST_SIMPLE("NULL => 1", 0, 1);
	LOCK_TEST_SIMPLE("Invalid => 1", 0xDEADBEEF, 1);
	LOCK_TEST_SIMPLE("Deleted => 1", mutex, 1);
	
	BASIC_SCHED_TEST("NULL",
		result = sceKernelLockMutex(0, 1, NULL);
	);
	BASIC_SCHED_TEST("Zero",
		result = sceKernelLockMutex(mutex2, 0, NULL);
	);
	BASIC_SCHED_TEST("Lock same",
		result = sceKernelLockMutex(mutex1, 1, NULL);
	);
	BASIC_SCHED_TEST("Lock other",
		result = sceKernelLockMutex(mutex2, 1, NULL);
	);
	BASIC_SCHED_TEST("Lock same with timeout",
		SceUInt timeout = 100;
		result = sceKernelLockMutex(mutex1, 1, &timeout);
	);
	BASIC_SCHED_TEST("Lock other with timeout",
		SceUInt timeout = 100;
		result = sceKernelLockMutex(mutex2, 1, &timeout);
	);

	SceUID exitThread = CREATE_SIMPLE_THREAD(exitFunc);
	mutex = sceKernelCreateMutex("lock", 0, 0, NULL);
	sceKernelStartThread(exitThread, sizeof(int), &mutex);
	sceKernelDelayThread(500);
	sceKernelLockMutex(mutex, 1, NULL);
	sceKernelDeleteMutex(mutex);
	schedf("Woke up after other thread exited.\n");

	return 0;
}