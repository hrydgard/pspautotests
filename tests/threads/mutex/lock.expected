Lock 0 => 0: Failed (800201BD)
Lock 0 => 1: Mutex: OK (size=56,name=lock,attr=00000000,init=0,current=1,lockThread=1,waiting=00000000)
Lock 0 => 2: Failed (800201BD)
Lock 0 => -1: Failed (800201BD)
Lock 1 => 1: Failed (800201C8)
Lock 0 => 2 (recursive): Mutex: OK (size=56,name=lock,attr=00000200,init=0,current=2,lockThread=1,waiting=00000000)
Lock 0 => -1 (recursive): Failed (800201BD)
Lock 1 => 1 (recursive): Mutex: OK (size=56,name=lock,attr=00000200,init=1,current=2,lockThread=1,waiting=00000000)
Lock 1 => INT_MAX - 1 (recursive): Mutex: OK (size=56,name=lock,attr=00000200,init=1,current=2147483647,lockThread=1,waiting=00000000)
Lock 1 => INT_MAX (recursive): Failed (800201C6)
Lock INT_MAX => INT_MAX (recursive): Failed (800201C6)
Lock 0 => 0: Failed (800201BD, 500ms left)
Lock 0 => 1: OK (500ms left)
Mutex: OK (size=56,name=lock,attr=00000000,init=0,current=1,lockThread=1,waiting=00000000)
Lock 0 => 2: Failed (800201BD, 500ms left)
Lock 0 => -1: Failed (800201BD, 500ms left)
Lock 1 => 1: Failed (800201C8, 500ms left)
Lock 0 => 2 (recursive): OK (500ms left)
Mutex: OK (size=56,name=lock,attr=00000200,init=0,current=2,lockThread=1,waiting=00000000)
Lock 0 => -1 (recursive): Failed (800201BD, 500ms left)
Lock 1 => 1 (recursive): OK (500ms left)
Mutex: OK (size=56,name=lock,attr=00000200,init=1,current=2,lockThread=1,waiting=00000000)
Zero timeout: OK (0ms left)
Mutex: OK (size=56,name=lock,attr=00000000,init=0,current=1,lockThread=1,waiting=00000000)
Locked 1 => 1: L2 L1 Failed (thread=800201A8, main=800201C8, 500ms left)
Locked 0 => 1: L1 L2 Failed (thread=00000000, main=800201A8, 0ms left)
Locked 1 => 1 (recursive): L2 L1 OK (thread=800201A8, 500ms left)
Locked 0 => 1 (recursive): L1 L2 Failed (thread=00000000, main=800201A8, 0ms left)
Locked zero timeout: L1 L2 Failed (thread=00000000, main=800201A8, 0ms left)
After delete: 800201B5
NULL => 0: Failed (800201C3)
NULL => 1: Failed (800201C3)
Invalid => 1: Failed (800201C3)
Deleted => 1: Failed (800201C3)
NULL: A1B1C2E1D1F2 (thread=800201B5, main=800201C3)
Zero: A1B1C2E1D1F2 (thread=800201B5, main=800201BD)
Lock same: A1B1C2E1D1F2 (thread=800201B5, main=800201C8)
Lock other: A1B1C2E1D1F2 (thread=800201B5, main=00000000)
Lock same with timeout: A1B1C2E1D1F2 (thread=800201B5, main=800201C8)
Lock other with timeout: A1B1C2E1D1F2 (thread=800201B5, main=00000000)
Now exiting: 00000000
Woke up after other thread exited.
